operator/pkg/remote/kyma.go

```diff
@@ -0,0 +1,184 @@
+	controlPlaneClient client.Client
+	runtimeClient      client.Client
+	controlPlaneKyma   *operatorv1alpha1.Kyma
+}
+
+func NewRemoteClient(ctx context.Context, controlPlaneClient client.Client, name, namespace string) (client.Client, error) {
+	cc := ClusterClient{
+		DefaultClient: controlPlaneClient,
+		Logger:        log.FromContext(ctx),
+	}
+
+	rc, err := cc.GetRestConfigFromSecret(ctx, name, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	remoteClient, err := client.New(rc, client.Options{Scheme: controlPlaneClient.Scheme()})
+	if err != nil {
+		return nil, err
+	}
+	return remoteClient, nil
+}
+
+func GetRemotelySyncedKyma(ctx context.Context, runtimeClient client.Client, key client.ObjectKey) (*operatorv1alpha1.Kyma, error) {
+	remoteKyma := &operatorv1alpha1.Kyma{}
+	if err := runtimeClient.Get(ctx, key, remoteKyma); err != nil {
+		return nil, err
+	}
+	return remoteKyma, nil
+}
+
+func DeleteRemotelySyncedKyma(ctx context.Context, controlPlaneClient client.Client, key client.ObjectKey) error {
+	runtimeClient, err := NewRemoteClient(ctx, controlPlaneClient, key.Name, key.Namespace)
+	if err != nil {
+		return err
+	}
+	remoteKyma, err := GetRemotelySyncedKyma(ctx, runtimeClient, key)
+	if err != nil {
+		return err
+	}
+	return runtimeClient.Delete(ctx, remoteKyma)
+}
+
+func RemoveFinalizerFromRemoteKyma(ctx context.Context, controlPlaneClient client.Client, key client.ObjectKey) error {
+	runtimeClient, err := NewRemoteClient(ctx, controlPlaneClient, key.Name, key.Namespace)
+	if err != nil {
+		return err
+	}
+	remoteKyma, err := GetRemotelySyncedKyma(ctx, runtimeClient, key)
+	if err != nil {
+		return err
+	}
+	controllerutil.RemoveFinalizer(remoteKyma, labels.Finalizer)
+	return runtimeClient.Update(ctx, remoteKyma)
+}
+
+func InitializeKymaSynchronizationContext(ctx context.Context, controlPlaneClient client.Client, controlPlaneKyma *operatorv1alpha1.Kyma) (*KymaSynchronizationContext, error) {
+	runtimeClient, err := NewRemoteClient(ctx, controlPlaneClient, controlPlaneKyma.Name, controlPlaneKyma.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	sync := &KymaSynchronizationContext{
+		controlPlaneClient: controlPlaneClient,
+		runtimeClient:      runtimeClient,
+		controlPlaneKyma:   controlPlaneKyma,
+	}
+
+	return sync, nil
+}
+
+func (c *KymaSynchronizationContext) CreateCRD(ctx context.Context) error {
+	crd := v1extensions.CustomResourceDefinition{}
+	err := c.controlPlaneClient.Get(ctx, client.ObjectKey{
+		// this object name is derived from the plural and is the default kustomize value for crd namings, if the CRD
+		// name changes, this also has to be adjusted here. We can think of making this configurable later
+		Name: fmt.Sprintf("%s.%s", operatorv1alpha1.KymaPlural, operatorv1alpha1.GroupVersion.Group),
+	}, &crd)
+	if err != nil {
+		return err
+	}
+	remoteCrd := v1extensions.CustomResourceDefinition{}
+	remoteCrd.Name = crd.Name
+	remoteCrd.Namespace = crd.Namespace
+	remoteCrd.Spec = crd.Spec
```

Comment:
If there is no CRD registered, there is a notRegisteredError and then returned. Where is the CRD created? I have a feeling I missed something..

Reply:
the crd is retrieved from the control plane since we can assume it is installed there, so it will fetch the crd from the control plane and then sync it into the SKR